File: Sorting and Searching/Nearest Smaller Values/NearestSmallerValues.cpp
#include<cstdio>
#include<algorithm>
#include<stack>
#include<vector>

using namespace std;

int main(){
    int n;
    scanf("%d", &n);
    vector<int> a(n);
    for(int i = 0; i < n; i++){
        scanf("%d", &a[i]);
    }
    stack<int> s;
    for(int i = 0; i < n; i++){
        while(!s.empty() && a[s.top()] >= a[i]){
            s.pop();
        }
        if(s.empty()){
            printf("0 ");
        }else{
            printf("%d ", s.top() + 1);
        }
        s.push(i);
    }
    printf("\n");
    fprintf(stderr, "Time: %fs\n", clock()/float(CLOCKS_PER_SEC));
    return 0;
}

File: Sorting and Searching/Nearest Smaller Values/NearestSmallerValues.py
import sys
from time import time

s = time()
input = sys.stdin.readline

n = int(input())
a = list(map(int, input().split()))

stack = []

for i in range(n):
    while stack and a[stack[-1]] >= a[i]:
        stack.pop()
    if stack:
        sys.stdout.write(str(stack[-1]+1) + ' ')
    else:
        sys.stdout.write('0 ')
    stack.append(i)
sys.stdout.write('\n')
sys.stderr.write("Time taken: "+str(time()-s)+" s\n")